src/command/subgraph/publish.rs

```diff
@@ -74,4 +126,84 @@ impl Publish {
             publish_response,
         })
     }
+
+    // output stream is optionally injected (defaulting to stdout) so we can test this function
+    pub fn handle_invalid_routing_url(
+        maybe_prompt_response: Option<bool>,
+        maybe_output: Option<&mut dyn std::io::Write>,
+    ) -> Option<RoverOutput> {
+        let default_output = &mut std::io::stdout();
+        let output = maybe_output.unwrap_or(default_output);
+        if let Some(prompt_response) = maybe_prompt_response {
+            if prompt_response {
+                None
+            } else {
+                writeln!(output, "Publish cancelled by user")
+                    .expect("Could not write to provided output stream");
+                Some(RoverOutput::EmptySuccess)
+            }
+        } else {
+            // if we're not in a tty, we didn't prompt. let's print a warning but
+            // publish anyway.
+            writeln!(
+                output,
+                "{} In a future major version of Rover, the `--allow-invalid-routing-url` flag will be required to publish a subgraph with an invalid routing URL in CI.",
+                Style::WarningPrefix.paint("WARN:")
+            ).expect("Could not write to provided output stream");
+            writeln!(
+                output,
+                "{} Found an invalid URL, but we can't prompt in a non-interactive environment. Publishing anyway.",
+                Style::WarningPrefix.paint("WARN:")
+            ).expect("Could not write to provided output stream");
+            None
+        }
+    }
+
+    pub fn prompt_for_publish() -> Option<bool> {
+        if !atty::is(atty::Stream::Stdout) {
```

Comment:
since we're prompting on `stderr` and reading from `stdin`, i think we should be running `atty::is` on both `stderr` and `stdin`. this probably doesn't match what we do elsewhere which isn't great. i think we should probably wrap these `atty` statements up directly in the `prompt` function suite, what do you think?

Reply:
I think you're right, but would rather open an issue / follow up for this since it introduces this trinary pattern of `Option<bool>` at other usage sites.